File: 1-2.py
example = """two1nine
eightwothree
abcone2threexyz
xtwone3four
4nineeightseven2
zoneight234
7pqrstsixteen"""

digit_names = [
    [1, "one"],
    [2, "two"],
    [3, "three"],
    [4, "four"],
    [5, "five"],
    [6, "six"],
    [7, "seven"],
    [8, "eight"],
    [9, "nine"],
]


def parse(line):
    for i, c in enumerate(line):
        if c.isdigit():
            yield int(c)

        parsed = next(
            (digit for [digit, name] in digit_names if line[i : i + len(name)] == name),
            None,
        )

        if parsed is not None:
            yield parsed

def numbers(input):
    for line in input.split("\n"):
        if len(line) == 0:
            continue

        digits = list(parse(line))

        yield digits[0] * 10 + digits[-1]


def solve(input):
    return sum(number for number in numbers(input))


if __name__ == "__main__":
    assert solve(example) == 281
    with open("1.txt", "r") as file:
        print(solve(file.read()))

File: 1-2.c
#include "prelude.h"

static const char *example =
	"two1nine\n"
	"eightwothree\n"
	"abcone2threexyz\n"
	"xtwone3four\n"
	"4nineeightseven2\n"
	"zoneight234\n"
	"7pqrstsixteen\n";

static const char *numbers[] = {
	"zero", "one", "two",   "three", "four",
	"five", "six", "seven", "eight", "nine",
};

static char getDigit(const char *text) {
	if (isdigit(*text)) return *text;
	for (unsigned char i = 0; i < ASZ(numbers); i++)
		if (!strncmp(text, numbers[i], strlen(numbers[i])))
			return '0' + i;
	return -1;
}

static int solve(const char *input) {
	int sum = 0;
	while (*input) {
		char digits[3] = {0};
		for (u32 i = 0; *input && *input != '\n'; input++) {
			const char digit = getDigit(input);
			if (digit < 0) continue;
			if (!i) digits[i++] = digit;
			digits[i] = digit;
		}
		sum += atoi(digits);
		input++;
	}
	return sum;
}

int main(void) {
	const int example_result = solve(example);
	if (example_result != 281) {
		printf("FAIL: expected 281 actual %d\n", example_result);
		return 1;
	}
	printf("Result: %d\n", solve(readToString("1.txt")));
	return 0;
}
